### PR TITLE
Display available tutorials

### LINKED ISSUE #223: `run_tutorial()` needs an example
Originally from @rpruim in https://github.com/rstudio/learnr/issues/208

### LINKED ISSUE #224: Ability to discover tutorials from a package
Originally from @rpruim in https://github.com/rstudio/learnr/issues/208

> missing name would yield a list of available tutorials in a package (much like the data() function does for data sets). Is there some other way to discover available tutorials?

Make a `available_tutorials()` function.  Run this function as the error message when a tutorial isn't found in `run_tutorial()`

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #224
Fixes #223 

``` r
learnr::available_tutorials()
#> Available tutorials:
#> * grader
#>   - grading-demo   : "Grading demo"
#>   - recording-demo : "Recording demo"
#>   - scoping-rules  : "What do exercise chunks have access to?"
#>   - setup-rules    : "learnr Setup Rules v 0.9.2"
#>   - solutions-demo : "Solutions demo"
#> * learnr
#>   - hello         : "Hello, Tutorial!"
#>   - question_type : "Question Types in Learnr"
#>   - slidy         : "Slidly demo"
learnr::available_tutorials("learnr")
#> Available tutorials:
#> * learnr
#>   - hello         : "Hello, Tutorial!"
#>   - question_type : "Question Types in Learnr"
#>   - slidy         : "Slidly demo"
learnr::run_tutorial()
#> Available tutorials:
#> * grader
#>   - grading-demo   : "Grading demo"
#>   - recording-demo : "Recording demo"
#>   - scoping-rules  : "What do exercise chunks have access to?"
#>   - setup-rules    : "learnr Setup Rules v 0.9.2"
#>   - solutions-demo : "Solutions demo"
#> * learnr
#>   - hello         : "Hello, Tutorial!"
#>   - question_type : "Question Types in Learnr"
#>   - slidy         : "Slidly demo"
learnr::run_tutorial(package = "learnr")
#> Available tutorials:
#> * learnr
#>   - hello         : "Hello, Tutorial!"
#>   - question_type : "Question Types in Learnr"
#>   - slidy         : "Slidly demo"
learnr::run_tutorial("not found")
#> Error: `package` must be provided if `name` is provided.
learnr::run_tutorial("not found", package = "learnr")
#> Error: Tutorial "not found" was not found in the "learnr" package.
#> Available tutorials:
#> * learnr
#>   - hello         : "Hello, Tutorial!"
#>   - question_type : "Question Types in Learnr"
#>   - slidy         : "Slidly demo"
```

<sup>Created on 2019-05-29 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

### CODE REVIEW

**Comment by wch** [comment]
May want to use `adist` to find the closest match, and say something like "Did you mean <closest match>?"

File: R/run.R [hunk 1]
```diff
@@ -15,23 +15,48 @@
...
+    # if any tutorial names are _close_ tell the user
+    adist_vals <- adist(possible_tutorials, name)
+    if (any(adist_vals <= 3)) {
+      best_match <- possible_tutorials[which.min(adist_vals)]
+      msg <- paste0(
+        msg, linesep,
+        "Did you mean \"", best_match, "\"?"
+      )
+    }
+    stop.(firstlinesep, msg, linesep, pkg_tutorials)
```
Comment (wch) [inline R/run.R]: Should be `stop`?

Comment (wch) [reply R/run.R]: Oh, nevermind, I see it now.

File: R/run.R [hunk 2]
```diff
@@ -45,6 +70,35 @@ run_tutorial <- function(name, package, shiny_args = NULL) {
...
 }
 
+#' @rdname run_tutorial
+#' @export
+available_tutorials <- function(package) {
+
+  if (!file.exists(
+    system.file(package = package)
+  )) {
+    stop.("No package found with name: \"", package, "\"")
```
Comment (wch) [inline R/run.R]: Would it be possible to show all packages with tutorials, sort of like how `data()` does it?

File: R/run.R [hunk 3]
```diff
@@ -15,23 +15,48 @@
...
 
   # validate that it's a direcotry
-  if (!utils::file_test("-d", tutorial_path))
-    stop("Tutorial ", name, " was not found in the ", package, " package.")
+  if (!utils::file_test("-d", tutorial_path)) {
+    firstlinesep <- "\t"
+    linesep <- "\n\t"
+    msg <- paste0("Tutorial \"", name, "\" was not found in the \"", package, "\" package.")
+    # if any tutorial names are _close_ tell the user
+    adist_vals <- adist(possible_tutorials, name)
```
Comment (wch) [inline R/run.R]: Maybe use `ignore.case=T`.

File: R/run.R (hunk 2 above)
Comment (schloerke) [reply R/run.R]: That would require traversing all packages.  There no "register" method for tutorials.  We can, just seems like a lot of computations.

File: R/run.R (hunk 3 above)
Comment (schloerke) [reply R/run.R]: I'm in the camp of "SaMe" does not equal "same".  The code below says they are equivalent.

```
❯❯ adist("SaMe", "same", ignore.case=TRUE)
     [,1]
[1,]    0
```

Ah, but it didn't exactly match before to get to this code path.  Ok. Fair!

Comment (schloerke) [reply R/run.R]: fixed

File: R/run.R (hunk 2 above)
Comment (schloerke) [reply R/run.R]: fixed.  happens when calling `available_tutorials()`

File: R/available_tutorials.R [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    ),
+    error = NULL
+  )
+}
+
+#' @return will return a list of `error` and `tutorials` which is a \code{data.frame} containing "package", "name", and "title".
+#' @noRd
+all_available_tutorials <- function() {
+  ret <- list()
+  all_pkgs <- base::library()$results[,"Package"]
```
Comment (wch) [inline R/available_tutorials.R]: It's little more idiomatic to use `installed.packages`.

File: R/available_tutorials.R [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  if (all(has_no_rmd)) {
+    return(an_error(
+      "No tutorial .Rmd files found for package: \"", package, "\""
+    ))
+  }
+
+  rmd_info <- rmd_info[!has_no_rmd]
+
+  tutorials <- do.call(rbind, rmd_info)
+  list(
```
Comment (wch) [inline R/available_tutorials.R]: A little more idiomatic to do:

```R
class(tutorials) <- c("learnr_available_tutorials", class(tutorials))
```

**Pull request merged**